File: Demos/07_FastAndSlow_ListCenter_demo.py
class Node:
    def __init__(self, value, next=None):
        self.val = value
        self.next = next

# Given the head of a LinkedList, returns the node at the middle of the list
def GetListCenter(head):
    slow = head
    fast = head
    while fast is not None and fast.next is not None:
        slow = slow.next
        fast = fast.next.next
    return slow

if __name__ == "__main__":
    # Insert elements
    head = Node(1)
    head.next = Node(2)
    head.next.next = Node(3)
    head.next.next.next = Node(4)
    head.next.next.next.next = Node(5)
    head.next.next.next.next.next = Node(6)
    head.next.next.next.next.next.next = Node(7)

    expectedOutput = 4
    output = GetListCenter(head).val

    print(output, expectedOutput, output == expectedOutput)

File: Demos/C++/07_FastAndSlow_ListCenter_demo.cpp
#include <iostream>

struct Node
{
    int value;
    Node* next;

    Node(int value): value(value), next(nullptr) {};
};

void PrintList(Node* head)
{
    Node* node = head;
    while (node != nullptr)
    {
        std::cout << node->value << " ";
        node = node->next;
    }
    std::cout << std::endl;
}

Node* FindListCenter(Node* head)
{
    Node* slow = head;
    Node* fast = head;

    while (fast != nullptr && fast->next != nullptr)
    {
        slow = slow->next;
        fast = fast->next->next;
    }

    return slow;
}

int main()
{
    // Insert elements
    Node* head;
    head = new Node(1);
    head->next = new Node(2);
    head->next->next = new Node(3);
    head->next->next->next = new Node(4);
    head->next->next->next->next = new Node(5);
    head->next->next->next->next->next = new Node(6);
    head->next->next->next->next->next->next = new Node(7);

    // Find list middle value
    PrintList(head);

    Node* middle = FindListCenter(head);

    std::cout << "\nMiddle value: " << middle->value << std::endl;
    return 0;
}